text/0000-output-options.md

```diff
@@ -0,0 +1,184 @@
+- Start Date: 2015-01-18
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Fix rustc’s handling of output (`-o`, `--out-dir` and `--emit`) options and their combinations.
+
+# Motivation
+
+rustc’s handling of various `-o`, `--out-dir`, and `--emit` command line option combinations is
+very inconsistent and sometimes unexpected. For example, the compiler will use filename provided
+via `-o` option for *some* of `--emit` targets, even though it emits a warning about filename being
+ignored. The document aims to propose rules governing interactions of options which influence
+compiler output.
+
+# Detailed design
+
+## `--out-dir`
+
+`--out-dir` is never ignored or adjusted, even in presence of other options that would usually
+conflict with `--out-dir`. If the option is provided and the directory does not exist yet, the
+compiler should create it. Defaults to `.`, otherwise known as current working directory or `$PWD`.
+Value of this option from now on will be referred to as `[out-dir]`.
+
+Note, that a requirement for the compiler to create the output directory if it didn’t exist yet is
+introduced. This is purely for convenience reasons: creating the output directory after a failure
+and restarting the program is exceedingly common nowadays.
+
+## `--emit` and `--crate-type`
+
+The two options influence what and how many files the compiler has to write to the filesystem. The
+common case compiler having to write one file only (e.g. `--emit` only has one value). In this case
+the path to output file shall be built using these templates:
+
+* `[out-dir]/[filename][.extension]` if `[filename]` was [inferred][inferred].
+* `[out-dir]/[filename]` otherwise;
+
+[inferred]: #-o-is-not-specified
+
+When there’s multiple files to output, each output file shall be written to path generated using
+`[out-dir]/[filename][.extension]` template.
+
+What `[filename]` resolves to is specified in [section about `-o`](#-o).
+
+The `[.extension]` is file extension specific to the emitted target. This is a list of extensions
+for each currently supported target:
+
+* asm – `.s`;
+* llvm-bc – `.bc`;
+* llvm-ir – `.ll`;
+* obj – `.o`;
+* dep-info – `.d`;
+* link – depends on `crate-type`:
+  * bin – no extension or `.exe`;
+  * lib – `.rlib`;
+  * rlib – `.rlib`;
+  * dylib – `.so`, `.dll` or `.dylib`;
+  * staticlib – `.a`.
```

Comment:
Does this interact with the changes to support MSVC?  If `rustc` is asked to produce a static library suitable for use with MSVC's `LINK`, it might need to use the `.lib` extension.  Obviously, this can be amended later, just thought I'd raise it now.